Add InfiniTime Vision document (#1196)
InfiniTimeOrg/InfiniTime: doc/InfiniTimeVision.md

> The purpose of this document is to steer efforts towards a common goal, and as such should be taken into consideration with all developments.

Reviewer: Let's put the `Vision` part as <h1> at the top of the document and make the other headers level 2 <h2>

```suggestion
# InfiniTime Vision

The purpose of this document is to steer efforts towards a common goal, and as such should be taken into consideration with all developments.
```

> # InfiniTime

Reviewer: make level 2 header

```suggestion
## InfiniTime
```

InfiniTime is a community built smartwatch firmware, providing freedom and privacy that's usually insufficient in wearable technology.

InfiniTime is not to be used for medical or other health tracking purposes.

> # Core Principles

Reviewer: make level 2 header

```suggestion
## Core Principles
```

> - KISS

Reviewer: just to be sure everyone knows what we mean
```suggestion
- KISS (Keep It Short and Simple)
```

- Reliability
- Battery efficiency
- Easy and simple navigation
- Behaviour should be predictable and easy to understand
- Prefer solid default experience over customization

> - Personalization is achieved through custom watchfaces. More options may be available through a companion app.

Reviewer: for markdown files it is a good custom to have one line per sentence. Then diffs are easier to read and one word change will only change one line, instead of possible a whole paragraph

```suggestion
- Personalization is achieved through custom watchfaces.
  More options may be available through a companion app.
```

- Use standard protocols and methods

> # Long term vision

Reviewer: make level 2 header

```suggestion
## Long term vision
```

The perfect version of InfiniTime would include:

- Capability to upload apps and watchfaces from a phone
- Only minimal features. Users would add the features they want.
- Ports to other devices
- Translations

> - Great documentation

Reviewer: user documentation or developer documentation or both?

Reviewer: I think yes, and we'll keep the code readable ourselfs ;)

# InfiniTime Vision

The purpose of this document is to steer efforts towards a common goal, and as such should be taken into consideration with all developments.

## InfiniTime

> InfiniTime is a community built smartwatch firmware, providing freedom and privacy that's usually insufficient in wearable technology.

Reviewer 2: Very minor grammatical issue, it should be community-built rather than community built.

> InfiniTime is a community-built smartwatch firmware, providing freedom and privacy that's usually insufficient in wearable technology.

Author: @Danimations has suggested to rephrase this as:
```suggestion
InfiniTime is a community-built firmware for the PineTime smartwatch. It offers freedom and privacy advantages unavailable to users of other proprietary wearable technology.
```

Perhaps saying it's only for the PineTime is too specific however, as there will be ports in the future.

Reviewer 2: I personally think that what is already there is mostly good, but that specifying that the other wearable technology it mentions is the proprietary kind would improve it.

Reviewer 3: Given your feedback, perhaps this is appropriate:

"InfiniTime is a community-built smartwatch firmware. It offers freedom and privacy advantages unavailable to users of proprietary wearable technology."

> InfiniTime is not to be used for medical or other health tracking purposes.

Reviewer: We don't intend InfiniTime to be used as medical device, because we can't (and don't want) to have to meet such a high standard and responsibility. But everyone can do what he/she wants with InfiniTime.

Or is the "not to be used" part some legal speak?
```suggestion
InfiniTime is not intended to be used for medical or other health tracking purposes.
```

Author: Of course everyone's free to do whatever regardless of what we say, but we want no part in that. I'd rather keep the statement strong and clear.

Reviewer: Good point!

## Core Principles

> - KISS (Keep It Short and Simple)

Reviewer 4: Imho this is vague. What about something like `write simple code, expand later if needed`. Since it's a mono-codebase it's useless to build supposedly ‶future-proof″ abstractions right away. We can refactor code once we have a clearer view about what kind of abstraction would be a great fit.

Reviewer 5: It's intentionally so. Also "Future-proof" abstractions are rarely short and simple.

Reviewer 4: @Avamander this is literally what I said.

> - KISS (Keep It Short and Simple)

Reviewer 6: I like this one!

- Reliability
- Battery efficiency
- Easy and simple navigation
- Behaviour should be predictable and easy to understand
- Prefer solid default experience over customization

> - Personalization is achieved through custom watchfaces.

Reviewer 4: Currently watchfaces are really hard to implement. Designers can't implement them because most of them can't program, and developers are predominantly terrible at UI/UX.

Once InfiniTime has external storage and this kind of stuff I'll probs work on a specification to make the creation of watchfaces easy & flexible. I'll harness my experience as a product engineer, and my close collaboration with designers in my previous work.

Reviewer 4: > Personalization is achieved through custom watchfaces.

Also, I believe that it contradicts the statement `capability to upload apps`.

Author: > Currently watchfaces are really hard to implement. Designers can't implement them because most of them can't program, and developers are predominantly terrible at UI/UX.

Makes sense. It would be great to get some collaboration happening between designers and developers.

> Also, I believe that it contradicts the statement capability to upload apps.

Hmm, how so?

Reviewer 4: > Makes sense. It would be great to get some collaboration happening between designers and developers.

Great! I'll work on mockups before implementing it so as to make sure we are on the same page. :)

> Hmm, how so?

I interpret this statement as exhaustive about how we can customize the OS. What do you think? It does not explicitly state “only” but many users and contributors won't pick the subtlety of the wording.

Do we plan to add other customizations than watchfaces and apps?

If there are not too many, I'd suggest making this list exhaustive.

Author: Ah I see what you mean. Here personalization means aesthetic, non-functional options, like changing fonts, certain colors etc. We've decided to limit aesthetic customization mainly to the watchface to keep things simple and try to provide a great user experience by default.

Many projects end up making a lot of things configurable, which quickly becomes overwhelming for users. We try to avoid adding too many options if possible, but not to the point of detriment.

  More options may be available through a companion app.

> - Use standard protocols and methods

Reviewer 4: Could we be more specific? For example BLE services with a link to a summary of the available ones. Maybe also explain the rational that justifies the burden of not just rolling out our custom easy-to-implement/debug solutions?

Reviewer 6: One of the goals of implementing standard and widely used protocols is to avoid re-inventing the wheel. Why spend time and efforts designing a protocol when we can just implement an existing one that fits our needs? Designing a protocol like DFU or file transfer protocol is not that easy. By using a well established one that is already used for similar use-case, we ensure that this protocol will probably work fine for us too.

Also, it makes the integration with other devices and apps a lot more easier. If a companion app already implements such a protocol, adding support for the PineTime will be virtually free. Why should we require companion apps developers to implement yet another protocol to set the time or to fetch heart rate values when they have already implement them multiple times?

In InfiniTime, we implement standard protocols from the BLE spec (GATT services : https://github.com/oesmith/gatt-xml).
For the BLE FS API, we implemented the BLE File Transfer Protocol from Adafruit (https://github.com/adafruit/Adafruit_CircuitPython_BLE_File_Transfer).
On the other hand, we couldn't find anything standard for the motion service, so we designed a custom one (https://github.com/InfiniTimeOrg/InfiniTime/blob/develop/doc/MotionService.md)

We choose the protocol on a case-by-case basis, but designing and implementing a new one from scratch should always be considered as a last resort.

## Long term vision

The perfect version of InfiniTime would include:

> - Capability to upload apps and watchfaces from a phone

Reviewer: Non-Blocking: Not just phone, but PC as well

```suggestion
- Capability to upload apps and watchfaces from a phone or computer
```

Reviewer 2: Perhaps it should be "using a companion app".

Author: How about this?
```suggestion
- Capability to sideload apps and watchfaces
```

> - Only minimal features.

Reviewer: ```suggestion
- Only minimal feature set in the flashed firmware.
```

> - Only a minimal feature set in the flashed firmware.

Reviewer 4: Doesn't this contradict the objective to `prefer solid default experience over customization`?
Imho we should remove non-essential stuff like pong, and move stuff like chimes to standalone apps. But useful features (find my phone, weather, calendar, etc.) should be available out of the box imho.

Author: It depends on the interpretation of minimal in this case. We intend to keep basic apps like for example stopwatch, timer and alarm. We'll need to consider which apps should be kept when the time comes, but that'll be a long time from now.

Reviewer 7: I think he idea of 'minimal' is like nothing too crazy-for example no big bowling app game but something like Snake would be acceptable

Reviewer 3: Bring on snake... if this is a serious prospect, I'd love to produce art for it.

> Users would add the features they want.

Reviewer: It seems to render fine without spaces, but in the text it is easer to read if indented
```suggestion
  Users would add the features they want.
```

- Ports to other devices

> - Translations

Reviewer 4: **Pitfall**: localizing apps is incredibly hard to do, and comes with a lot of unexpected challenges. PineTime requires users to be tech-savy — so they can overwhelmingly understand basic English. Is support for other languages worth it before a commercial pre-installed noob-friendly watch gets released?

Author: There isn't a plan currently to implement translations, but it's certainly something that would be great to have someday, hence I think long term vision is the correct place for this.
